hail/python/hailtop/hailctl/config/cli.py

```diff
@@ -42,29 +142,29 @@ def get_section_key_path(parameter: str) -> Tuple[str, str, Tuple[str, ...]]:
     sys.exit(1)
 
 
+def complete_config_variable(incomplete: str):
+    for name, help_text, _ in config_variables():
+        if name.startswith(incomplete):
+            completion_item = (name, help_text)
+            yield completion_item
+
+
 @app.command()
-def set(parameter: str, value: str):
+def set(parameter: Ann[str, Arg(default=..., help="Configuration variable to set", autocompletion=complete_config_variable)], value: str):
     '''Set a Hail configuration parameter.'''
-    from hailtop.aiotools.router_fs import RouterAsyncFS  # pylint: disable=import-outside-toplevel
     from hailtop.config import get_user_config, get_user_config_path  # pylint: disable=import-outside-toplevel
 
     config = get_user_config()
     config_file = get_user_config_path()
     section, key, path = get_section_key_path(parameter)
 
-    validations = {
-        ('batch', 'bucket'): (
-            lambda x: re.fullmatch(r'[^:/\s]+', x) is not None,
-            'should be valid Google Bucket identifier, with no gs:// prefix',
-        ),
-        ('batch', 'remote_tmpdir'): (
-            RouterAsyncFS.valid_url,
-            'should be valid cloud storage URI such as gs://my-bucket/batch-tmp/',
-        ),
-        ('email',): (lambda x: re.fullmatch(r'.+@.+', x) is not None, 'should be valid email address'),
-    }
+    validations = {name: validation for name, _, validation in config_variables()}
 
-    validation_func, msg = validations.get(path, (lambda _: True, ''))  # type: ignore
+    if parameter not in validations:
+        print(f"Error: unknown parameter {parameter}", file=sys.stderr)
+        sys.exit(1)
+
+    validation_func, msg = validations.get(parameter, (lambda _: True, ''))  # type: ignore
```

Comment:
```suggestion
    validation_func, msg = validations[parameter]
```

We've now guaranteed that `parameter` will be present in the dictionary.

Reply:
Right, I was worried that if I missed some configuration in the code base or something we add later on, then we've basically made this version of hailctl unusable. So basically, my goal was just to make what we have more helpful rather than stricter. So the real question is whether lines 163-165 are correct.

Reply:
I'm happy to pose this question to the larger team, but in my opinion it is a bug to accept arbitrary config values not a feature because if I'm allowed to set it I would expect it to do something. If we agree it's a bug, we are within our rights to revoke the ability to set arbitrary config values.

Regarding versions being unusable, that's an interesting question. I guess that would happen if someone does somewhere in the code `configuration_of(some_new_config)` without ever adding validations for it? And that would only make it through our tests if we only ever use the environment variable override instead of using `hailctl config set` which to be honest sounds very plausible. I suppose that for that problem I would propose the solution that `configuration_of(` should not accept some arbitrary string, but a variant of an `Enum` such that we can enforce in a test that every variant of that enum is represented here, such that you  *can't* introduce configuration without also adding support for it in `hailctl`. This would also give a lot more confidence that this PR caught all usages of `configuration_of`.

Reply:
I forgot to address this comment.